File: proxy/nodejs/src/db/adapters/oracle.js
'use strict';
const oracledb = require('oracledb');

const _pools = new Map();  // configKey -> Pool
let _thickInitDone = false;

// Thick 모드 초기화 (Oracle Instant Client 필요).
// Thin 모드는 Oracle DB 12.1+ 만 지원하므로, 구버전 DB 연결 시 Thick 모드가 필요하다.
// initOracleClient 는 프로세스당 한 번만 호출 가능하므로 플래그로 보호한다.
// Windows에서 libDir 파라미터는 DLL 로딩 방식 차이로 불안정하므로 시스템 PATH에 등록된 경로를 사용한다.
function initThickMode() {
  if (_thickInitDone) return;
  _thickInitDone = true;
  try {
    oracledb.initOracleClient();
  } catch (_) {
    // Oracle Instant Client가 시스템 PATH에 없음 → Thin 모드 유지 (Oracle DB 12.1+ 만 가능)
  }
}

function configKey(config) {
  return JSON.stringify({
    host: config.host,
    port: config.port || 1521,
    database: config.database,
    user: config.username
  });
}

async function getPool(config) {
  initThickMode();
  const key = configKey(config);
  if (_pools.has(key)) return _pools.get(key);

  const connectString = `${config.host}:${config.port || 1521}/${config.database}`;
  const pool = await oracledb.createPool({
    user:          config.username,
    password:      config.password,
    connectString,
    poolMin:       0,
    poolMax:       10,
    poolTimeout:   30,
    queueTimeout:  10000
  });
  _pools.set(key, pool);
  return pool;
}

async function execute(config, sql) {
  const pool = await getPool(config);
  let conn;
  try {
    conn = await pool.getConnection();
    const result = await conn.execute(sql, [], {
      outFormat: oracledb.OUT_FORMAT_OBJECT,
      fetchArraySize: 100
    });
    const rows = result.rows || [];
    const fields = result.metaData ? result.metaData.map(m => m.name.toLowerCase()) : [];
    return {
      rows: rows.map(row => {
        const out = {};
        for (const [k, v] of Object.entries(row)) out[k.toLowerCase()] = v;
        return out;
      }),
      rowCount: result.rowsAffected != null ? result.rowsAffected : rows.length,
      fields
    };
  } finally {
    if (conn) { try { await conn.close(); } catch (_) {} }
  }
}

async function executeParams(config, sql, params) {
  const pool = await getPool(config);
  let conn;
  try {
    conn = await pool.getConnection();
    // oracledb positional 바인딩: :1, :2, ... 형식
    const binds = params.map(v => v);
    const result = await conn.execute(sql, binds, {
      outFormat: oracledb.OUT_FORMAT_OBJECT,
      fetchArraySize: 100,
      autoCommit: true
    });
    const rows = result.rows || [];
    const fields = result.metaData ? result.metaData.map(m => m.name.toLowerCase()) : [];
    return {
      rows: rows.map(row => {
        const out = {};
        for (const [k, v] of Object.entries(row)) out[k.toLowerCase()] = v;
        return out;
      }),
      rowCount: result.rowsAffected != null ? result.rowsAffected : rows.length,
      fields
    };
  } finally {
    if (conn) { try { await conn.close(); } catch (_) {} }
  }
}

async function test(config) {
  const result = await execute(config, 'SELECT 1 AS ok FROM DUAL');
  return result.rows.length > 0;
}

async function closePool(key = null) {
  if (key) {
    const pool = _pools.get(key);
    if (pool) { try { await pool.close(0); } catch (_) {} _pools.delete(key); }
  } else {
    for (const pool of _pools.values()) { try { await pool.close(0); } catch (_) {} }
    _pools.clear();
  }
}

module.exports = { execute, executeParams, test, closePool };

File: proxy/python/db/adapters/oracle.py
import asyncio
import oracledb

_pools: dict = {}  # config_key -> oracledb.ConnectionPool
_thick_init_done = False


def _init_thick_mode(client_lib_dir: str = None) -> None:
    global _thick_init_done
    if _thick_init_done:
        return
    _thick_init_done = True
    try:
        oracledb.init_oracle_client(lib_dir=client_lib_dir or None)
    except Exception:
        pass


def _config_key(config: dict) -> str:
    return f"{config['host']}:{config.get('port', 1521)}:{config['database']}:{config['username']}"


def _get_pool(config: dict):
    _init_thick_mode(config.get("clientLibDir"))
    key = _config_key(config)
    pool = _pools.get(key)
    if pool:
        return pool
    connect_string = f"{config['host']}:{config.get('port', 1521)}/{config['database']}"
    pool = oracledb.create_pool(
        user=config["username"],
        password=config["password"],
        dsn=connect_string,
        min=0,
        max=10,
        increment=1,
        timeout=30,
    )
    _pools[key] = pool
    return pool


async def execute(config: dict, sql: str) -> dict:
    loop = asyncio.get_running_loop()
    return await loop.run_in_executor(None, _execute_sync, config, sql)


def _execute_sync(config: dict, sql: str) -> dict:
    pool = _get_pool(config)
    conn = pool.acquire()
    try:
        cursor = conn.cursor()
        cursor.execute(sql)
        if cursor.description:
            fields = [col[0].lower() for col in cursor.description]
            rows = [dict(zip(fields, row)) for row in cursor.fetchall()]
            return {"rows": rows, "rowCount": len(rows), "fields": fields}
        # 결과셋 없음(INSERT/UPDATE/DELETE/DDL) → 커밋. oracledb 기본 autocommit=False.
        conn.commit()
        return {"rows": [], "rowCount": cursor.rowcount or 0, "fields": []}
    finally:
        pool.release(conn)


def _execute_params_sync(config: dict, sql: str, params) -> dict:
    """파라미터 바인딩 실행 (:1, :2, ... 플레이스홀더). SELECT는 rows 반환, DML은 rowCount 반환 + 커밋."""
    pool = _get_pool(config)
    conn = pool.acquire()
    try:
        cursor = conn.cursor()
        cursor.execute(sql, params)
        if cursor.description:
            fields = [col[0].lower() for col in cursor.description]
            rows = [dict(zip(fields, row)) for row in cursor.fetchall()]
            return {"rows": rows, "rowCount": len(rows), "fields": fields}
        # DML — 항상 커밋 (oracledb 기본 autocommit=False)
        conn.commit()
        return {"rows": [], "rowCount": cursor.rowcount or 0, "fields": []}
    finally:
        pool.release(conn)


async def execute_params(config: dict, sql: str, params) -> dict:
    """파라미터 바인딩 비동기 실행 (executor 래핑)."""
    loop = asyncio.get_running_loop()
    return await loop.run_in_executor(None, _execute_params_sync, config, sql, params)


async def test(config: dict) -> bool:
    result = await execute(config, "SELECT 1 AS ok FROM DUAL")
    return len(result["rows"]) > 0


async def close_pool(key: str = None) -> None:
    global _pools
    if key:
        pool = _pools.pop(key, None)
        if pool:
            try:
                pool.close()
            except Exception:
                pass
    else:
        for p in list(_pools.values()):
            try:
                p.close()
            except Exception:
                pass
        _pools = {}
